Add service module and test for Taskserver

aszlig opened the pull request:
This picks up on #7729 and particularly on #7771, where this branch is based on (plus my earlier work in matthiasbeyer/nixpkgs#3). However, it's without the commits from @jgeerds, because I already had refactored the service module back then (which didn't get pushed to #7771).

The implementation works and the test passes, but there are still a few issues to fix, such as:
- [x] Add a way to specify certificate generation options, especially for >2048 bit keys.
- [x] Less verbose certificate creation, because right now it spams the journal all over the place during the startup.
- [x] Make sure we don't interfere with user-set certificates in the helper tool.
- [x] Check whether CA exists in the helper tool to avoid race conditions.
- [x] Document usage of `nixos-taskserver` in the module documentation.
- [x] Allow to remove and suspend users, groups and organisations.
- [x] Introduce an `extraConfig` option (maybe even call it `config` and make it an `attrsOf`?).
- [x] Improve the systemd service definition (things such as `execReload`, `chroot`, `PrivateTmp`, etc.. maybe even socket activation?).
- [x] Rewrite `nixos-taskserver` in one programming language only.

I'd like to apologize to @matthiasbeyer in particular for not getting back to this earlier.

Cc: @matthiasbeyer, @jgeerds and also @nbp, @lethalman, @jagajaga, @bjornfor, @Lassulus who commented/reviewed #7771

matthiasbeyer committed:
Add services file for taskwarrior server service

matthiasbeyer committed:
Add taskserver to module-list

matthiasbeyer committed:
taskd service: Add initialization script

matthiasbeyer committed:
Add taskserver to ids.nix

aszlig committed:
nixos/ids: Rename uid and add gid for "taskd"

I'm renaming the attribute name for uid, because the user name is called
"taskd" so we should really use the same name for it.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Unify taskd user and group

The service doesn't start with the "taskd" user being present, so we
really should add it. And while at it, it really makes sense to add a
default group as well.

I'm using a check for the user/group name as well, to allow the
taskserver to be run as an existing user.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Remove options for log/pidFile

We're aiming for a proper integration into systemd/journald, so we
really don't want zillions of separate log files flying around in our
system.

Same as with the pidFile. The latter is only needed for taskdctl, which
is a SysV-style initscript and all of its functionality plus a lot more
is handled by systemd already.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Improve module options

The descriptions for the options previously seem to be from the
taskdrc(5) manual page. So in cases where they didn't make sense for us
I changed the wording a bit (for example for client.deny we don't have a
"comma-separated list".

Also, I've reordered things a bit for consistency (type, default,
example and then description) and add missing types, examples and
docbook tags.

Options that are not used by default now have a null value, so that we
can generate a configuration file out of all the options defined for the
module.

The dataDir default value is now /var/lib/taskserver, because it doesn't
make sense to put just yet another empty subdirectory in it and "data"
doesn't quite make sense anyway, because it also contains the
configuration file as well.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Refactor module for CA creation

Now the service starts up if only the services.taskserver.enable option
is set to true.

We now also have three systemd services (started in this order):

 * taskserver-init: For creating the necessary data directory and also
                    includes a refecence to the configuration file in
                    the Nix store.
 * taskserver-ca:   Only enabled if none of the server.key, server.cert,
                    server.crl and caCert options are set, so we can
                    allow for certificates that are issued by another
                    CA.
                    This service creates a new CA key+certificate and a
                    server key+certificate and signs the latter using
                    the CA key.
                    The permissions of these keys/certs are set quite
                    strictly to allow only the root user to sign
                    certificates.
 * taskserver:      The main Taskserver service which just starts taskd.

We now also log to stdout and thus to the journal.

Of course, there are still a few problems left to solve, for instance:

 * The CA currently only signs the server certificates, so it's
   only usable for clients if the server doesn't validate client certs
   (which is kinda pointless).
 * Using "taskd <command>" is currently still a bit awkward to use, so
   we need to properly wrap it in environment.systemPackages to set the
   dataDir by default.
 * There are still a few configuration options left to include, for
   example the "trust" option.
 * We might want to introduce an extraConfig option.
 * It might be useful to allow for declarative configuration of
   organisations and users, especially when it comes to creating client
   certificates.
 * The right signal has to be sent for the taskserver service to reload
   properly.
 * Currently the CA and server certificates are created using
   server.host as the common name and doesn't set additional certificate
   information. This could be improved by adding options that explicitly
   set that information.

As for the config file, we might need to patch taskd to allow for
setting not only --data but also a --cfgfile, which then omits the
${dataDir}/config file. We can still use the "include" directive from
the file specified using --cfgfile in order to chainload
${dataDir}/config.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Properly indent CA config lines

No change in functionality, but it's easier to read when properly
indented.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Add trust option to config file

The server starts up without that option anyway, but it complains about
its value not being set. As we probably want to have access to that
configuration value anyway, let's expose this via the NixOS module as
well.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Use types.str instead of string

The "string" option type has been deprecated since a long time
(800f9c2), so let's not use it here.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Introduce a new fqdn option

Using just the host for the common name *and* for listening on the port
is quite a bad idea if you want to listen on something like :: or an
internal IP address which is proxied/tunneled to the outside.

Hence this separates host and fqdn.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Fix generating server cert

We were generating a self-signed certificate for the server so far,
which we obviously don't want.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Add an option for organisations

We want to declaratively specify users and organisations, so let's add
another module option "organisations", which allows us to specify users,
groups and of course organisations.

The implementation of this is not yet done and this is just to feed the
boilerplate.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Add a nixos-taskdctl command

It's a helper for NixOS systems to make it easier to handle CA
certificate signing, similar to what taskd provides but comes preseeded
with the values from the system configuration.

The tool is very limited at the moment and only allows to *add*
organisations, users and groups. Deletion and suspension however is much
simpler to implement, because we don't need to handle certificate
signing.

Another limitation is that we don't take into account whether
certificates and keys are already set in the system configuration and if
they're set it will fail spectacularly.

For passing the commands to the taskd command, we're using a small C
program which does setuid() and setgid() to the Taskserver user and
group, because runuser(1) needs PAM (quite pointless if you're already
root) and su(1) doesn't allow for setting the group and setgid()s to the
default group of the user, so it even doesn't work in conjunction with
sg(1).

In summary, we now have a shiny nixos-taskdctl command, which lets us do
things like:

nixos-taskdctl add-org NixOS
nixos-taskdctl add-user NixOS alice
nixos-taskdctl export-user NixOS alice

The last command writes a series of shell commands to stdout, which then
can be imported on the client by piping it into a shell as well as doing
it for example via SSH:

ssh root@server nixos-taskdctl export-user NixOS alice | sh

Of course, in terms of security we need to improve this even further so
that we generate the private key on the client and just send a CSR to
the server so that we don't need to push any secrets over the wire.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Use nixos-taskdctl in preStart

Finally, this is where we declaratively set up our organisations and
users/groups, which looks like this in the system configuration:

services.taskserver.organisations.NixOS.users = [ "alice" "bob" ];

This automatically sets up "alice" and "bob" for the "NixOS"
organisation, generates the required client keys and signs it via the
CA.

However, we still need to use nixos-taskdctl export-user in order to
import these certificates on the client.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/tests: Add a test for the Taskserver service

A small test which checks whether tasks can be synced using the
Taskserver.

It doesn't test group functionality because I suspect that they're not
yet implemented upstream. I haven't done an in-depth check on that but I
couldn't find a method of linking groups to users yet so I guess this
will get in with one of the text releases of Taskwarrior/Taskserver.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Factor out nixos-taskdctl

With a cluttered up module source it's really a pain to navigate through
it, so it's a good idea to put it into another file.

No changes in functionality here, just splitting up the files and fixing
references.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Rename nixos-taskdctl

Using nixos-taskserver is more verbose but less cryptic and I think it
fits the purpose better because it can't be confused to be a wrapper
around the taskdctl command from the upstream project as
nixos-taskserver shares no commonalities with it.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

jagajaga commented:
:astonished: Great job!

nbp commented on nixos/modules/services/misc/taskserver/helper-tool.nix:
```diff
@@ -0,0 +1,324 @@
+  mkCStr = val: "\"${lib.escape ["\\" "\""] val}\"";
+
+  taskdUser = let
+    runUser = pkgs.writeText "runuser.c" ''
```
I do not understand the need for this C file.  Why couldn't we split the systemd job in 2, one which does the preliminary setup, and the other one which start the service with a specific user/group?

aszlig replied:
The C file is only needed for `nixos-taskwarrior` to setuid()/setgid(), see the message of 227229653a414fc6fe6c387ff3e01ef506267da9 for the reasons.

matthiasbeyer replied:
Would it be beneficial to move this source to an extra file? At least I'd like to have a comment _in_ the source why this is needed and how this is used.

matthiasbeyer replied:
Ah, moving it to an own file would be more complex than this way as you generate some variables from nix code. nvm! :smile:

aszlig replied:
Thought about this during sleep and I think the tool should indeed be written in one language, because adding C and using Nix to avoid limitations and/or boilerplate in the shell isn't very good if you want things to stay consistent and readable. So going to add this to the todo above.

nbp commented:
Can you explain a bit more what is the `nixos-taskserver` binary doing, and why we need an extra tool?  Shouldn't we contribute such tool upstream instead?  Why is it implemented in 3 different languages: Nix, Bash and C?

aszlig commented:
The `nixos-taskserver` is there for setting up the CA needed for Taskserver in conjunction with declarative users. At first I added it solely for specifying users/groups declaratively but thought it would be a good idea to expose it to `systemPackages` as well, so if someone doesn't want to do that he/she doesn't need to copy & paste together a bunch of different store paths.

aszlig commented:
And no, it doesn't make sense to contribute this upstream, except if they'd want to adopt the NixOS module system :-)

matthiasbeyer commented:
@aszlig Wow, thank you very much! This is amazing! I will review the code now.

matthiasbeyer commented on nixos/modules/services/misc/taskserver/default.nix:
```diff
@@ -0,0 +1,430 @@
+          type = types.int;
+          default = 53589;
+          description = ''
+            Port number of the Taskserver.
```
We might add a short note that this port has to be opened in the firewall, or is this done by the service automatically?

aszlig replied:
The listening host is `localhost` by default, so opening a port wouldn't work here anyway, because access to the loopback interface is unfiltered. However it could make sense add `networking.firewall.allowedTCPPorts` if the host is not `localhost`.

aszlig committed:
nixos/taskserver: Fix type for client.{allow,deny}

We already document that we allow special values such as "all" and
"none", but the type doesn't represent that. So let's use an enum in
conjuction with a loeOf type so that this becomes clear.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Fix type/description for ciphers

Referring to the GnuTLS documentation isn't very nice if the user has to
use a search engine to find that documentation. So let's directly link
to it.

The type was "str" before, but it's actually a colon-separated string,
so if we set options in multiple modules, the result is one concatenated
string.

I know there is types.envVar, which does the same as separatedString ":"
but I found that it could confuse the reader of the Taskserver module.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Rename client.{allow,deny}

These values match against the client IDs only, so let's rename it to
something that actually reflects that. Having client.cert in the same
namespace also could lead to confusion, because the client.cert setting
is for the *debugging* client only.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Add module documentation

It's not by any means exhaustive, but we're still going to change the
implementation, so let's just use this as a starting point.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Remove client.cert option

The option is solely for debugging purposes (particularly the unit tests
of the project itself) and doesn't make sense to include it in the NixOS
module options.

If people want to use this, we might want to introduce another option so
that we can insert arbitrary configuration lines.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Rewrite helper-tool in Python

In the comments of the pull request @nbp wrote:

"Why is it implemented in 3 different languages: Nix, Bash and C?"

And he's right, it doesn't make sense, because we were using C as a
runuser replacement and used Nix to generate the shellscript
boilerplates.

Writing this in Python gets rid of all of this and we also don't need
the boilerplate as well, because we're using Click to handle all the
command line stuff.

Note that this currently is a 1:1 implementation of what we had before.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/tests/taskserver: Test imperative users

As the nixos-taskserver command can also be used to imperatively manage
users, we need to test this as well.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Rename server.{host,port}

Having an option called services.taskserver.server.host is quite
confusing because we already have "server" in the service name, so let's
first get rid of the listening options before we rename the rest of the
options in that .server attribute.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Rename .server options to .pki

After moving out the PKI-unrelated options, let's name this a bit more
appropriate, so we can finally get rid of the taskserver.server thing.

This also moves taskserver.caCert to taskserver.pki.caCert, because that
clearly belongs to the PKI options.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Move .pki.fqdn to .fqdn

It's not necessarily related to the PKI options, because this is also
used for setting the server address on the Taskwarrior client.

So if someone doesn't have his/her own certificates from another CA, all
options that need to be adjusted are in .pki. And if someone doesn't
want to bother with getting certificates from another CA, (s)he just
doesn't set anything in .pki.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Move .trust out of .pki

This is clearly a server configuration option and has nothing to do with
certificate creation and signing, so let's move it away from the .pki
namespace.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Improve doc for PKI options

The improvement here is just that we're adding a big <note/> here so
that users of these options are aware that whenever they're setting one
of these the certificates and keys are _not_ created automatically.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Rename .pki options

We're now using .pki.server.* and .pki.ca.* so that it's entirely clear
what these keys/certificates are for. For example we had just .pki.key
before, which doesn't really tell very much about what it's for except
if you look at the option description.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Set server.crl for automatic CA

Currently, we don't handle this yet, but let's set it so that we cover
all the options.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/doc: Allow refs from options to the manual

My first attempt to do this was to just use a conditional <refsection/>
in order to not create exact references in the manpage but create the
reference in the HTML manual, as suggested by @edolstra on IRC.

Later I went on to use <olink/> to reference sections of the manual, but
in order to do that, we need to overhaul how we generate the manual and
manpages.

So, that's where we are now:

There is a new derivation called "manual-olinkdb", which is the olinkdb
for the HTML manual, which in turn creates the olinkdb.xml file and the
manual.db. The former contains the targetdoc references and the latter
the specific targetptr elements.

The reason why I included the olinkdb.xml verbatim is that first of all
the DTD is dependent on the Docbook XSL sources and the references
within the olinkdb.xml entities are relative to the current directory.

So using a store path for that would end up searching for the manual.db
directly in /nix/store/manual.db.

Unfortunately, the <olinks/> that end up in the output file are
relative, so for example if you're clicking on one of these within the
PDF, the URL is searched in the current directory.

However, the sections from the olink's text are still valid, so we could
use an alternative URL for that in the future.

The manual doesn't contain any links, so even referencing the relative
URL shouldn't do any harm.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>
Cc: @edolstra

aszlig committed:
nixos/taskserver: Link to manual within .enable

With <olink/> support in place, we can now reference the Taskserver
section within the NixOS manual, so that users reading the manpage of
configuration.nix(5) won't miss this information.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig commented:
Cc: @edolstra because of 1d77dcaed37ab47bfe2d90711c01b475a514ff25.

matthiasbeyer commented:
Holy crap, the more code you introduce here the more do I recognize how few things I know about nixos service writing! Thank you so much for doing this, @aszlig !

aszlig committed:
nixos/taskserver/helper: Factor out program logic

The Click functions really are for the command line and should be solely
used for that.

What I have in mind is that instead of that crappy --service-helper
argument, we should really have a new subcommand that is expecting JSON
which is directly coming from the services.taskserver.organisations
module option.

That way we can decrease even more boilerplate and we can also ensure
that organisations, users and groups get properly deleted if they're
removed from the NixOS configuration.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Handle declarative conf via JSON

We now no longer have the stupid --service-helper option, which silences
messages about already existing organisations, users or groups.

Instead of that option, we now have a new subcommand called
"process-json", which accepts a JSON file directly from the specified
NixOS module options and creates/deletes the users accordingly.

Note that this still has a two issues left to solve in this area:

 * Deletion is not supported yet.
 * If a user is created imperatively, the next run of process-json will
   delete it once deletion is supported.

So we need to implement deletion and a way to mark organisations, users
and groups as "imperatively managed".

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Constrain server cert perms

It doesn't do much harm to make the server certificate world readable,
because even though it's not accessible anymore via the file system,
someone can still get it by simply doing a TLS handshake with the
server.

So this is solely for consistency.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Generate a cert revocation list

If we want to revoke client certificates and want the server to actually
notice the revocation, we need to have a valid certificate revocation
list.

Right now the expiration_days is set to 10 years, but that's merely to
actually get certtool to actually generate the CRL without trying to
prompt for user input.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Add a command to reload service

Unfortunately we don't have a better way to check whether the reload has
been done successfully, but at least we now *can* reload it without
figuring out the exact signal to send to the process.

Note that on reload, Taskserver will not reload the CRL file. For that
to work, a full restart needs to be done.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver/helper: Implement deletion

Now we finally can delete organisations, groups and users along with
certificate revocation. The new subtests now make sure that the client
certificate is also revoked (both when removing the whole organisation
and just a single user).

If we use the imperative way to add and delete users, we have to restart
the Taskserver in order for the CRL to be effective.

However, by using the declarative configuration we now get this for
free, because removing a user will also restart the service and thus its
client certificate will end up in the CRL.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/tests/taskserver: Make tests less noisy

We were putting the whole output of "nixos-taskserver export-user" from
the server to the respective client and on every such operation the
whole output was shown again in the test log.

Now we're *only* showing these details whenever a user import fails on
the client.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Silence certtool everywhere

We only print the output whenever there is an error, otherwise let's
shut it up because it only shows information the user can gather through
other means. For example by invoking certtool manually, or by just
looking at private key files (the whole blurb it's outputting is in
there as well).

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Don't change imperative users

Whenever the nixos-taskserver tool was invoked manually for creating an
organisation/group/user we now add an empty file called .imperative to
the data directory.

During the preStart of the Taskserver service, we use process-json which
in turn now checks whether those .imperative files exist and if so, it
doesn't do anything with it.

This should now ensure that whenever there is a manually created user,
it doesn't get killed off by the declarative configuration in case it
shouldn't exist within that configuration.

In addition, we also add a small subtest to check whether this is
happening or not and fail if the imperatively created user got deleted
by process-json.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Allow to specify expiration/bits

At least this should allow for some customisation of how the
certificates and keys are created. We now have two sub-namespaces within
PKI so it should be more clear which options you have to set if you want
to either manage your own CA or let the module create it automatically.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Introduce an extraConfig option

This is simply to add configuration lines to the generated configuration
file. The reason why I didn't went for an attribute set is that the
taskdrc file format doesn't map very well on Nix attributes, for example
the following can be set in taskdrc:

server = somestring
server.key = anotherstring

In order to use a Nix attribute set for that, it would be way too
complicated, for example if we want to represent the mentioned example
we'd have to do something like this:

{ server._top = somestring;
  server.key = anotherstring;
}

Of course, this would work as well but nothing is more simple than just
appending raw strings.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Setup CA before main service

We need to explicitly make sure the CA is created before we actually
launch the main Taskserver service in order to avoid race conditions
where the preStart phase of the main service could possibly corrupt
certificates if it would be started in parallel.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver/helper: Assert CA existence

We want to make sure that the helper tool won't work if the automatic CA
wasn't properly set up. This not only avoids race conditions if the tool
is started before the actual service is running but it also fails if
something during CA setup has failed so the user can investigate what
went wrong.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Reorder into one mkMerge

No changes in functionality but rather just restructuring the module
definitions to be one mkMerge, which now uses mkIf from the top-level
scope of the CA initialization service so we can better abstract
additional options we might need there.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Set allowedTCPPorts accordingly

As suggested by @matthiasbeyer:

"We might add a short note that this port has to be opened in the
firewall, or is this done by the service automatically?"

This commit now adds the listenPort to
networking.firewall.allowedTCPPorts as soon as the listenHost is not
"localhost".

In addition to that, this is now also documented in the listenHost
option declaration and I have removed disabling of the firewall from the
VM test.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver/helper: Improve CLI subcommands

Try to match the subcommands to act more like the subcommands from the
taskd binary and also add a subcommand to list groups.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Finish module documentation

Apart from the options manual, this should cover the basics for setting
up a Taskserver. I am not a native speaker so this can and (probably)
should be improved, especially the wording/grammar.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Set up service namespaces

The Taskserver doesn't need access to the full /dev nor does it need a
shared /tmp. In addition, the initialisation services don't need network
access, so let's constrain them to the loopback device.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Improve service dependencies

Using requiredBy is a bad idea for the initialisation units, because
whenever the Taskserver service is restarted the initialisation units
get restarted as well.

Also, make sure taskserver-init.service will be ordered *before*
taskserver.service.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver: Restart service on failure

This is the recommended way for long-running services and ensures that
Taskserver will keep running until it has been stopped manually.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig commented:
Okay, so far this is finished and it's ready for full review, further improvements can still be made after merging.

aszlig committed:
nixos/taskserver: Fix wrong option doc references

The options client.allow and client.deny are gone since the commit
8b793d1916387c67f8eeb137789b1b41a1f94537, so let's fix that.

No feature changes, only fixes the descriptions of allowedClientIDs and
disallowedClientIDs.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/taskserver/helper: Fix docstring of add_user

We have already revamped the CLI subcommands in commit
e2383b84f88e0e7d35f6a3a846b54c69e3bee6ee.

This was just an artifact that was left because of this.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

edolstra commented:
What's the impact on the build time of the manual?

Regarding the PDF issue, we should probably just drop the PDF, I don't think anybody will miss it.

aszlig commented:
@edolstra:

Without the commit:
  `nix-build --check nixos/release.nix -A manual.x86_64-linux  0.73s user 0.08s system 5% cpu 13.574 total`
With the commit:
  `nix-build --check nixos/release.nix -A manual.x86_64-linux  0.75s user 0.09s system 6% cpu 13.556 total`

edolstra commented:
And how much *including* the olink derivation? Because I assume that will need to be regenerated every time Nixpkgs changes.

aszlig commented:
@edolstra: This is just for the olink derivation:
`nix-store --check -r  0.01s user 0.00s system 0% cpu 4.037 total`
So it's around +4s build time.

aszlig commented:
Other than that, @edolstra, @nbp and @matthiasbeyer: Any issues still unresolved? If not, please give a thumbs up and I'll merge.

edolstra commented:
What does a link to the manual in a manpage look like btw?

aszlig commented:
@edolstra: In case of Taskserver it is: `More instructions about NixOS in conjuction with Taskserver can be found in the NixOS manual at Chapter 15, Taskserver.`

nbp commented on nixos/doc/manual/default.nix:
```diff
@@ -72,6 +73,63 @@ let
       </toc>
     '';
 
+  manualXsltprocOptions = toString [
```
These changes sounds like an issue that should be fixed any way, can you open another pull request to get these changes merged independently of the taskserver modification.

aszlig replied:
I'll revert the commit here and open it as a new PR separately. The build of the manual won't fail if this PR is merged but it will issue an ugly warning about unresolved olinks. The manpage will contain something like `More instructions about NixOS in conjuction with Taskserver can be found in the NixOS manual at ???`, so I guess it shouldn't do much harm except for the missing reference in the Taskserver option description.

aszlig replied:
Reverted in c36d6e59647794aa1c4dbb2dcd7e5242d9c5b6b8 and brought back in #14700.

nbp commented on nixos/modules/services/misc/taskserver/doc.xml:
```diff
@@ -0,0 +1,144 @@
+  services.taskserver.organisations.NixOS.users = [ "alice" ];
+}
+</screen>
+      This creates an organisation called <literal>NixOS</literal> with the user
```
Choose a better organization name in this example, such that it is less confusing.  Maybe something like `my-company`

nbp commented:
This looks good to me, apart from the previous remarks on the modifications.
:+1:

aszlig committed:
nixos/taskserver/doc: Improve example org name

Suggested by @nbp:

"Choose a better organization name in this example, such that it is less
confusing. Maybe something like my-company"

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

aszlig committed:
nixos/doc: Revert allowing olinks from options

This reverts commit 1d77dcaed37ab47bfe2d90711c01b475a514ff25.

It will be reintroduced along with #14700 as a separate branch, as
suggested by @nbp.

I added this to this branch because I thought it was a necessary
dependency, but it turns out that the build of the manual/manpages still
succeeds and merely prints a warning like this:

warning: failed to load external entity "olinkdb.xml"
Olink error: could not open target database 'olinkdb.xml'.
Error: unresolved olink: targetdoc/targetptr = 'manual/module-taskserver'.

The olink itself will be replaced by "???", so users looking at the
description of the option in question will still see the reference to
the NixOS manual, like this:

More instructions about NixOS in conjuction with Taskserver can be found
in the NixOS manual at ???.

Signed-off-by: aszlig <aszlig@redmoonstudios.org>

matthiasbeyer commented:
I'm not good enough with Nix and the Nixpkgs infrastructure to have a real opinion whether this is okay or not... I do not have concerns, so feel free to go on without my "ACK". You have my blessings anyways! :smile:
